alf/algorithms/mcts_algorithm.py

```diff
@@ -505,19 +646,59 @@ def _search(self, trees, to_plays):
         children_index = trees.children_index
         best_child_index = trees.best_child_index
         game_over = trees.game_over
+        psims = self._num_parallel_sims
+        parallel = self._parallel
         search_paths = []
         B = trees.B
         nodes = trees.root_indices
-        path_lengths = torch.ones_like(B)
+        calc_i = False
+        if parallel:
+            B = B.unsqueeze(-1)
+            nodes = trees.root_indices.unsqueeze(-1).expand(-1, psims)
+            if self._search_with_exploration_policy:
+                i = (torch.arange(psims), )
+            else:
+                calc_i = True
+            if self._is_two_player_game:
+                to_plays = to_plays.unsqueeze(-1).expand_as(nodes)
+        else:
+            i = ()
+        # [B] or [B, psims]
+        path_lengths = torch.ones_like(nodes)
         search_paths = [nodes]
         if game_over is not None:
             done = game_over[B, nodes]
         else:
-            done = torch.zeros_like(B, dtype=torch.bool)
-        while not torch.all(done):
+            done = torch.zeros_like(nodes, dtype=torch.bool)
+        depth = 0
+        while not torch.all(done) and depth < self._max_depth:
+            depth += 1
+            if calc_i:
+                # If several nodes are same, we should select actions from different
+                # entries in best_child_index. So we need to find out which entry
+                # of best_child_index a node should use. For search_with_exploration_policy,
+                # the entries in best_child_index are sampled independently from
+                # the calculated policy, so it is ok to use i'th best_child_index
+                # for i-th node. But for UCB based selection, we want to use
+                # the top entries in best_child_index when ever they are not used.
+                #
+                # For x>y, i[b, x, y] indicates where nodes[b, x] and nodes[b, y] are same
+                # For x<=y, i[b, x, y] is 0.
+                # For example, suppose nodes[0] = [2,3,2,3,3], i[0] will be
+                # [[0, 0, 0, 0, 0],
+                #  [0, 0, 0, 0, 0],
+                #  [1, 0, 0, 0, 0],
+                #  [0, 1, 0, 0, 0],
+                #  [0, 1, 0, 1, 0]]
+                # [B, psims, psims]. `None` stands for unsqueezing at that dim.
+                i = (nodes[:, :, None] == nodes[:, None, :]).tril(diagonal=-1)
+                # With the above example, i[0] is [0, 0, 1, 1, 2]
+                # [B, psims]
```

Comment:
A further example on how this is affecting the `torch.where` below would be very helpful. I have to admit that this alone takes quite some time to digest.